### PR TITLE
Building templates

### PR DESCRIPTION
**Pull request by teubert**
- Updated cmakelists to also build the templates. 
- Fixed any issues in the templates so they build

### CODE REVIEW

**Comment by teubert** [comment]
@jason-watkins note: This includes what I had in the other pull request (I had to include it to be able to build)

File: templates/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+set(HEADERS
+	EmptyCommunicator.h
+	EmptyModel.h
+	EmptyObserver.h
+        EmptyPredictor.h
```
Comment (jason-watkins) [inline templates/CMakeLists.txt]: Indentation is off here (probably mixed tabs and spaces).

File: templates/EmptyModel.cpp [hunk 2]
```diff
@@ -30,74 +28,35 @@ EmptyModel::EmptyModel(const ConfigMap& configMap) : EmptyModel::EmptyModel() {
...
-    // x[0] = a + adot*dt;
-    // x[1] = b + bdot*dt;
-    // ...
-
-    // Add process noise
-    // x[0] += dt*n[0];
-    // x[1] += dt*n[1];
-    // ...
+Model::state_type EmptyModel::stateEqn(double t,
+                                         const Model::state_type& x,
```
Comment (jason-watkins) [inline templates/EmptyModel.cpp]: More inconsistent whitespace

File: templates/EmptyModel.cpp [hunk 3]
```diff
@@ -30,74 +28,35 @@ EmptyModel::EmptyModel(const ConfigMap& configMap) : EmptyModel::EmptyModel() {
...
-    // ...
-
-    // Add noise
-    // z[0] += n[0];
-    // ...
+Model::output_type EmptyModel::outputEqn(double t,
+                                         const Model::state_type& x,
+                                         const Model::input_type& u,
+                                         const Model::noise_type& n) const {
+    Model::output_type output;
```
Comment (jason-watkins) [inline templates/EmptyModel.cpp]: These types aren't automatically sized, so this line should be `auto output = getOutputVector();` to get an output vector of the correct size. Maybe we should delete the default constructor of `DynamicArray` to make sure this mistake isn't possible.

File: templates/EmptyObserver.cpp [hunk 4]
```diff
@@ -19,72 +19,46 @@
...
+    
+    EmptyObserver::EmptyObserver(const Model* m, const Matrix Q, const Matrix R) : Observer(m) {
+        
     }
     
     // Initialize function (required by Observer interface)
-    void EmptyObserver::initialize(const double t0, const std::vector<double> & x0,
-                                           const std::vector<double> & u0) {
+    void EmptyObserver::initialize(const double t0,
+                                                const Model::state_type & x0,
```
Comment (jason-watkins) [inline templates/EmptyObserver.cpp]: More whitespace

File: templates/EmptyObserver.cpp [hunk 5]
```diff
@@ -19,72 +19,46 @@
...
+                                                const Model::input_type & u0) {
         log.WriteLine(LOG_DEBUG, MODULE_NAME, "Initializing");
+        Model::state_type initialized_state;
```
Comment (jason-watkins) [inline templates/EmptyObserver.cpp]: `auto initialized_state = getStateVector();`

File: templates/EmptyObserver.h [hunk 6]
```diff
@@ -23,37 +23,66 @@
...
+        
         /** @brief Initialize UKF
         *   @param t0 Initial time
         *   @param x0 Initial state vector
         *   @param u0 Initial input vector
         **/
-        void initialize(const double t0, const std::vector<double> & x0,
-            const std::vector<double> & u0);
+        void initialize(const double t0,
+                                     const Model::state_type & x0,
```
Comment (jason-watkins) [inline templates/EmptyObserver.h]: Whitespace

File: templates/EmptyPrognosticsModel.cpp [hunk 7]
```diff
@@ -24,116 +30,65 @@ EmptyPrognosticsModel::EmptyPrognosticsModel(const ConfigMap & configMap) : Empt
...
-    // State equations
-    // double adot = a + b;
-    // double bdot = c;
-    // ...
+Model::state_type EmptyPrognosticsModel::stateEqn(double t,
+                                       const Model::state_type& x,
+                                       const Model::input_type& u,
+                                       const Model::noise_type& n,
+                                       double dt) const {
+    Model::state_type new_state;
```
Comment (jason-watkins) [inline templates/EmptyPrognosticsModel.cpp]: `auto new_state = getStateVector();`

File: templates/EmptyPrognosticsModel.cpp [hunk 8]
```diff
@@ -24,116 +30,65 @@ EmptyPrognosticsModel::EmptyPrognosticsModel(const ConfigMap & configMap) : Empt
...
-    // ...
-    
-    // Extract inputs
-    // double c = u[0];
-    // ...
+Model::output_type EmptyPrognosticsModel::outputEqn(double t,
+                                         const Model::state_type& x,
+                                         const Model::input_type& u,
+                                         const Model::noise_type& n) const {
+    Model::output_type output;
```
Comment (jason-watkins) [inline templates/EmptyPrognosticsModel.cpp]: `auto output = getOutputVector();`

File: templates/EmptyPrognosticsModel.cpp [hunk 9]
```diff
@@ -24,116 +30,65 @@ EmptyPrognosticsModel::EmptyPrognosticsModel(const ConfigMap & configMap) : Empt
...
-    // Compute based on voltage, so use output equation to get voltage
-    std::vector<double> z(2);
-    std::vector<double> zeroNoise(8);
-    outputEqn(t, x, u, zeroNoise, z);
-
-    // Determine if voltage (second element in z) is below VEOD threshold
-    return z[1] <= parameters.VEOD;
+// Initialize state, given initial inputs and outputs
+Model::state_type EmptyPrognosticsModel::initialize(const Model::input_type& u, const Model::output_type& z) const {
+    Model::state_type initialized_state;
```
Comment (jason-watkins) [inline templates/EmptyPrognosticsModel.cpp]: `auto initialized_state = getStateVector();`

File: templates/EmptyPrognosticsModel.cpp [hunk 10]
```diff
@@ -24,116 +30,65 @@ EmptyPrognosticsModel::EmptyPrognosticsModel(const ConfigMap & configMap) : Empt
...
-    // double b = x[1];
-    // ...
-    
-    // Extract inputs
-    // double c = u[0];
-    // ...
+Model::input_type EmptyPrognosticsModel::inputEqn(double t,
+                                     const std::vector<double>& params,
+                                     const std::vector<double>& loadEstimate) const {
+    Model::input_type input;
```
Comment (jason-watkins) [inline templates/EmptyPrognosticsModel.cpp]: `auto input = getInputVector();`

**Pull request merged**